src/Microsoft.Health.Dicom.Core/Features/CustomTag/CustomTagService.cs

```diff
@@ -0,0 +1,84 @@
+﻿// -------------------------------------------------------------------------------------------------
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License (MIT). See LICENSE in the repo root for license information.
+// -------------------------------------------------------------------------------------------------
+
+using System;
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+using EnsureThat;
+using Microsoft.Extensions.Logging;
+using Microsoft.Health.Dicom.Core.Messages.CustomTag;
+
+namespace Microsoft.Health.Dicom.Core.Features.CustomTag
+{
+    public class CustomTagService : ICustomTagService
+    {
+        private readonly ICustomTagStore _customTagStore;
+        private readonly IReindexJob _reindexJob;
+        private readonly ICustomTagEntryValidator _customTagEntryValidator;
+        private readonly ILogger<CustomTagService> _logger;
+
+        public CustomTagService(ICustomTagStore customTagStore, IReindexJob reindexJob, ICustomTagEntryValidator customTagEntryValidator, ILogger<CustomTagService> logger)
+        {
+            EnsureArg.IsNotNull(customTagStore, nameof(customTagStore));
+            EnsureArg.IsNotNull(reindexJob, nameof(reindexJob));
+            EnsureArg.IsNotNull(customTagEntryValidator, nameof(customTagEntryValidator));
+            EnsureArg.IsNotNull(logger, nameof(logger));
+
+            _customTagStore = customTagStore;
+            _reindexJob = reindexJob;
+            _customTagEntryValidator = customTagEntryValidator;
+            _logger = logger;
+        }
+
+        public async Task<AddCustomTagResponse> AddCustomTagAsync(IEnumerable<CustomTagEntry> customTags, CancellationToken cancellationToken = default)
+        {
+            // Validate input
+            await _customTagEntryValidator.ValidateCustomTagsAsync(customTags, cancellationToken);
+
+            HashSet<long> addedTagKeys = new HashSet<long>();
+            foreach (var tag in customTags)
+            {
+                try
+                {
+                    long key = await _customTagStore.AddCustomTagAsync(tag.Path, tag.VR, tag.Level, CustomTagStatus.Reindexing);
```

Comment:
Would it make more sense to do the "add/record a reindex job/update the status" as a database transaction/stored procedure such that we might be able to avoid weird scenarios where we've added portions of the required information but then failed out.

Reply:
once we have the job framework, we will know how to handle this.
For now, added note in TODO.